lib/src/annotate.rs

```diff
@@ -0,0 +1,335 @@
+            self.local_line_map.insert(new_commit_id.clone(), new_map);
+        }
+    }
+
+    fn swap_full_commit_id(&mut self, old_commit_id: &CommitId, new_commit_id: &CommitId) {
+        let old_commit_map = self.local_line_map.remove(old_commit_id);
+        if old_commit_map.is_none() {
+            return;
+        }
+        let old_commit_map = old_commit_map.unwrap();
+        self.local_line_map
+            .insert(new_commit_id.clone(), old_commit_map);
+    }
+
+    fn drain_remaining_for_commit_id(&mut self, commit_id: &CommitId) {
+        let remaining_lines = self.local_line_map.remove(commit_id);
+        if remaining_lines.is_none() {
+            return;
+        }
+        let remaining_lines = remaining_lines.unwrap();
+        for (_, original_line_number) in remaining_lines {
+            self.original_line_map
+                .insert(original_line_number, commit_id.clone());
+        }
+    }
+
+    fn get_next_commit(&self) -> Option<CommitId> {
+        return self.local_line_map.keys().next().cloned();
+    }
+
+    fn convert_to_results(self, original_contents: &[u8]) -> AnnotateResults {
+        let original_content_lines: Vec<String> =
+            original_contents.lines().map(|s| s.unwrap()).collect();
+        let mut result_lines = Vec::new();
+        for (idx, line) in original_content_lines.into_iter().enumerate() {
+            result_lines.push((self.original_line_map.get(&idx).unwrap().clone(), line));
+        }
+        AnnotateResults {
+            file_annotations: result_lines,
+        }
+    }
+}
+
+/// Get line by line annotations for a specific file path in the repo.
+pub fn get_annotation_for_file(
+    file_name: &str,
+    repo: &ReadonlyRepo,
+    starting_commit_id: &CommitId,
+) -> Result<AnnotateResults, AnnotateError> {
+    let file_path = RepoPathBuf::from_relative_path(file_name);
+    if file_path.is_err() {
+        return Err(AnnotateError::FileNotFound(file_name.to_string()));
+    }
+    let file_path: RepoPathBuf = file_path.unwrap();
+    let store = repo.store();
+
+    let current_commit = store.get_commit(starting_commit_id)?;
+    let current_tree = current_commit.tree()?;
+    let original_file_id = get_file_id(current_commit.id(), &current_tree, &file_path)?;
+    if original_file_id.is_none() {
+        return Err(AnnotateError::FileNotFound(file_name.to_string()));
+    }
+    let original_contents =
+        get_file_contents(repo.store(), &file_path, &original_file_id.unwrap())?;
+    let num_lines = original_contents.lines().count();
+    let mut partial_results = PartialResults::new(starting_commit_id, num_lines);
+
+    process_commits(repo, current_commit.id(), &file_path, &mut partial_results)?;
+
+    Ok(partial_results.convert_to_results(&original_contents))
+}
+
+fn process_commits(
+    repo: &ReadonlyRepo,
+    starting_commit_id: &CommitId,
+    file_name: &RepoPath,
+    results: &mut PartialResults,
+) -> Result<(), AnnotateError> {
```

Comment:
Nit: Why is this not `fn process_commits(...) -> Result<PartialResults, AnnotateError)`?

Reply:
I can, but is there a specific reason for why you think it's better?

Reply:
Isn't linear control flow easier to read instead of having output parameters?

Reply:
IMO, it's not too bad to have `& mut` parameters if the function receives a partially-processed object and processes it further, but then  it should be the first argument since modifying this object is the most important thing the function does.

Even in this case, I would often reach for a function that would take a `PartialResults` (not a reference) and return a `PartialResults`. But (perhaps unlike Philip), I don't see this as much better, I don't have a clear rule about how to choose.

Separately, re-designing the logic to decrease the need for such functions is a good thing, if possible. I'm not immediately sure if that could be done here.

A big no-no for me (which I think is not the case here) is if a function takes an empty object as &mut and uses &mut purely as an output.